torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -380,11 +398,25 @@ def __init__(
                 cpu_offload=cpu_offload,
                 backward_prefetch=backward_prefetch,
                 mixed_precision=mixed_precision,
+                param_init_fn=param_init_fn,
             )
 
         self.process_group = process_group or _get_default_group()
         self.rank = self.process_group.rank()
         self.world_size = self.process_group.size()
+
+        # We assume that the module is initialized on the meta device
+        # if the user has specified param_init_fn. TODO: investigate
+        # a way to more robustly determine if the module is on the meta
+        # device or not.
+        is_meta_module = param_init_fn is not None
+        if is_meta_module:
+            # Initialize parameters according to the user lambda.
+            assert callable(param_init_fn), (
```

Comment:
Wonder whether we can check module is meta or not based on module or parameter' device attributes? also make param_init_fn default to be None? so that if module is a meta device, we can detect it internally and call reset_parameter for it in default if param_init_fn is None. if reset_parameter is not available or param_init_fn is not provided, throw errors. 
Basically move the definition codes of _init_with_reset_params in the test into FSDP internal codes?

Reply:
Sounds good, updated based on this feedback.

Now we check for is_meta or is_fake (if torchdistx is avail) and have default init functions for both of these scenarios. User can still pass in `param_init_fn` if they want.